### ISSUE TITLE
Player sequencing needs to be defined better

### ISSUE THREAD

**Issue by birtles** [w3c/web-animations#62]
Does it even work with the current model for transitions etc.? Does it get reset when you cancel and re-start a player?

**Comment by birtles** [comment]
So I think there are two issues here: **layering** and **priority/sequence**.

In terms of **layering** we have the following arrangement:

1. Transitions apply to the transition level of the cascade which overrides the animation level (see [Cascading](http://dev.w3.org/csswg/css-cascade-3/#cascade-origin)).
2. CSS Animations and script-generated animations apply to the animation level of the cascade.

In terms of **priority/sequence**, we have the following constraints:
* Transitions have the lowest priority. If there is a CSS Animation targeting the same property that is active at the same time, the transition doesn't affect the property. This is despite the transitions otherwise being applied to the transitions level of the cascade that is higher up! See [Application of transitions](http://dev.w3.org/csswg/css-transitions/#application).
* Within transitions, there's no concept of priority because each transition, by definition, targets a different property. Longhands get expanded and if a property is repeated, only the last one is used (see [transition-property](http://dev.w3.org/csswg/css-transitions/#transition-property-property)). However _we will need to define the order of transitions_ returned from <code>getAnimations</code>.
* Within CSS Animations, the order of the computed value of <code>animation-name</code> goes from lower priority to higher priority ("If multiple animations are attempting to modify the same property, then the animation closest to the end of the list of names wins." See [animation-name](http://dev.w3.org/csswg/css-animations/#animation-name))
* Within CSS Animations, if we have a make a change from <code>animation: anim 2s</code> to <code>animation: anim 4s, anim 6s</code> we have to decide what happens. In Chrome, I believe the one animation gets dropped. In Firefox we produce both animations and the original animation gets mapped to the second instance in the new list. That is, the above acts like a *prepend*.
** I believe having two animations is more intuitive from an authoring point-of-view.
** It also has practical uses (I have to dig up the link for this but I was able to solve one author's problem by relying on having two instances of the same animation).
** As for why the example above prepends vs appending, dbaron thought this made more sense. I disagreed at first, but eventually came to agree. Some context in [Mozilla bug 1037316](https://bugzilla.mozilla.org/show_bug.cgi?id=1037316).

I think the constraints on CSS Animations are sufficiently complicated that we won't be able to interleave script-generated animations with CSS Animations. At one point I'm pretty sure I analyzed this more thoroughly and convinced myself it was pretty much impossible unless we relax the constraints somehow.

As a result, I think we will probably end up making script-generated animations have priority over CSS Animations. So I think we get a prioritization like the following (going lowest priority to highest):

1. CSS Transitions
    a. Sort by animation sequence number and define how this works for style change events where several transitions begin (in the CSS Transitions spec)
2. CSS Animations
     a. Sort by animation-name &larr; this works for <code>Element.getAnimations()</code> but not for <code>AnimationTimeline.getAnimations()</code>. Perhaps the sequence beyond that is just document order?
3. Script-generated animations
     a. Sort by animation sequence number

That's a rough start at outlining this. I think it suggests we need hooks on the spec for animation sources which have a both priority and a target cascade level.

**Comment by birtles** [comment]
Also, I was wondering the other day if animation sequence should be based on when the animation gets attached to a timeline but that doesn't work since animations without a timeline can still affect an element and hence need a sequence number.

**Comment by birtles** [comment]
Another issue is the behavior of script-generated <code>CSSTransition</code> and <code>CSSAnimation</code> objects.

* For <code>CSSTransition</code> objects, we could say they always target the transitions level of the cascade? Might be useful if you want to write an animation that clobbers all others regardless of its start time? i.e. **levelling** is a property of the *type* of animation. But how about prioritization? That should probably be a property of the animation *source* (i.e. transitions markup, animations markup, API call). (There might be some complications here regarding how addition works, but so far I think it might be ok.)
* <code>CSSAnimation</code> objects they should still target the animation level of the cascade. What about their prioritization though? If they are script-generated objects they won't appear in the <code>animation-name</code> list so we can't use that. We could probably do a couple of things here:
   1.  Say all script-generated <code>CSSAnimation</code> objects either come before or after (whichever proves more useful) the markup-generated <code>CSSAnimation</code> objects and are sorted by animation sequence number.
   2.  Say all script-generated <code>CSSAnimation</code> objects are identical to regular <code>Animation</code> objects as far as prioritization goes.

   I think the second option is better. But that begs the question, why would you use a <code>CSSAnimation</code> object at all? In part it doesn't matter. We'll add a constructor to the <code>CSSAnimation</code> interface simply for the sake of there being less magic. But, we'll probably also make it take a CSSKeyframesRule parameter so using <code>CSSAnimation</code> could be a useful way of re-using a CSSKeyframesRule.

**Comment by dstockwell** [comment]
> In Chrome, I believe the one animation gets dropped.

This was not intentional, tracked here: https://code.google.com/p/chromium/issues/detail?id=487092

> As for why the example above prepends vs appending, dbaron thought this made more sense. I disagreed at first, but eventually came to agree. Some context in Mozilla bug 1037316.

Could you summarize that? I couldn't quite see the reasoning in the thread.

> I think the second option is better. But that begs the question, why would you use a CSSAnimation object at all? In part it doesn't matter. We'll add a constructor to the CSSAnimation interface simply for the sake of there being less magic. But, we'll probably also make it take a CSSKeyframesRule parameter so using CSSAnimation could be a useful way of re-using a CSSKeyframesRule.

This all seems quite reasonable to me.

**Comment by birtles** [comment]
So as a first pass, I think the prioritization could be something like:

1. Animations generated by transitions markup.

    (The following sorting within transitions is mostly for the sake of making the return value from <code>getAnimations()</code> consistent across implementations. Typically it has no effect on composite order since transition properties don't overlap but they *could* if the <code>CSSTransition</code> object is updated by script.)

    1. Sort by when the transition got triggered. This would be the animation sequence number except that we process style changes in batches so we need to define an order within those transitions that got triggered in the same batch (which we might then reflect back into the sequence number).
    1. For transitions that got triggered at the same time sort by document order of the element that *generated* the transition (recognizing that a generated <code>CSSTransition</code> may subsequently have a different effect attached that targets a different element).
    1. For transitions that got triggered at the same time on the same element, sort by transition property name.
1. Animations generated by animations markup
    1. Sort by document order of the element the *generated* the animation
    1. For animations generated by the same element, sort by the order in which they appear in the <code>animation-name</code> property.
1. Animations generated by script
     1. Sort by animation sequence number

Note that part of the complication here is also that in future <code>Animation</code>s will be able to target multiple elements through groups. It's tempting to first sort by document order then return the set of <code>Animation</code> objects per element/pseudo but that doesn't work once we have animations targeting multiple element.

I think 1.i might be unnecessary. Or at least it's clearly inconsistent with CSS Animations which is partly because of the extra constraints put on CSS Animations.

**Comment by birtles** [comment]
Note that when it comes to compositing, it might be as simple as, "take the animated value produced by the last animation in the stack and if it is a <code>CSSTransition</code> object, stick it in the transition level of the cascade, otherwise stick it in the animation level." (But I have a feeling I'll later discover that doesn't work and we need more special handling for transitions.)

**Comment by birtles** [comment]
>> As for why the example above prepends vs appending, dbaron thought this made more sense. I disagreed at first, but eventually came to agree. Some context in Mozilla bug 1037316.

> Could you summarize that? I couldn't quite see the reasoning in the thread.

In part it was simply because it was unspecced ("Until the CSS spec says otherwise, I'd prefer to stick with the preserving-last behavior") but also because matching from the end will mean there's no jarring restart if the new animation has a shorter duration than the old. Matching from the end may be more consistent with other CSS features, I'm not sure. If you think otherwise we can definitely revisit that behaviour.

**Comment by birtles** [comment]
Sent mail to try and clarify one of the edge cases here: https://lists.w3.org/Archives/Public/public-fx/2015AprJun/0054.html

**Comment by birtles** [comment]
> I think 1.i might be unnecessary. Or at least it's clearly inconsistent with CSS Animations which is partly because of the extra constraints put on CSS Animations.

Actually, I think we do need this. Without this the return value of <code>elem.getAnimations()</code> is not very intuitive. e.g.

```js
elem.style.transition = 'all 1s';
getComputedStyle(elem).transform;
elem.style.transform = 'translateX(100px');
elem.getAnimations(); // Returns [ CSSTransition { transitionProperty: 'transform' ... } ]
elem.style.opacity = '0';
elem.getAnimations(); // I think you expect to get the transform transition followed
                      // by the opacity transition here
```

**Comment by birtles** [comment]
Made these changes:
https://github.com/w3c/web-animations/compare/103c2ccb98bd...ce4e853582067

Need to add the appropriate bits to CSS Animations and CSS Transitions, however.

**Comment by birtles** [comment]
CSS Animations:
https://rawgit.com/shans/csswg-drafts/animations-2/css-animations-2/Overview.html#animation-priority

**Comment by birtles** [comment]
CSS Transitions:
https://rawgit.com/shans/csswg-drafts/animations-2/css-transitions-2/Overview.html#animation-priority

**Comment by dbaron** [comment]
> For CSSTransition objects, we could say they always target the transitions level of the cascade? Might be useful if you want to write an animation that clobbers all others regardless of its start time? i.e. levelling is a property of the type of animation. But how about prioritization? That should probably be a property of the animation source (i.e. transitions markup, animations markup, API call). (There might be some complications here regarding how addition works, but so far I think it might be ok.)

I'm not sure where you ended up on this issue, but I don't think we want script-generated CSSTransition objects to have the "override everything" behavior.  It's ok for UA-generated CSS transitions to be able to override everything since they're transitions between values that were generated in other ways, so we know that both values won in the cascade.  But part of the goal of the CSS cascade is that authors aren't necessarily allowed to override everything.

**Comment by birtles** [comment]
> I'm not sure where you ended up on this issue, but I don't think we want script-generated CSSTransition objects to have the "override everything" behavior. It's ok for UA-generated CSS transitions to be able to override everything since they're transitions between values that were generated in other ways, so we know that both values won in the cascade. But part of the goal of the CSS cascade is that authors aren't necessarily allowed to override everything.

Currently script-generated CSSTransition objects apply to the transitions level of the cascade. I've just checked this and it currently isn't specced properly, however. I'll file a separate issue to fix this.

I think we want to apply the same behavior to script-generated CSSTransition objects and CSSTransition objects that are no longer associated with markup. e.g.
```js
elem.style.left = '100px';
elem.style.transition = 'left 10s';
getComputedStyle(elem).transitionProperty;
elem.style.left = '200px';
var transition = elem.getAnimations()[0]; // getAnimations flushes style so this works
elem.style.transition = ''; // This effectively calls cancel() on |transition|
transition.play();
```

In this example script keeps the CSSTransition object alive but it is no longer tied to markup. In this case, I presume <code>transition</code> should no longer target the transitions-level of the cascade.
